request-manager/spec/sequence-check.go

```diff
@@ -0,0 +1,161 @@
+		}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type OptionalArgsHaveDefaultsSequenceCheck struct{}
+
+/* Optional args must have defaults. */
+func (check OptionalArgsHaveDefaultsSequenceCheck) CheckSequence(sequence SequenceSpec) error {
+	for _, arg := range sequence.Args.Optional {
+		if arg.Default == nil {
+			return MissingValueError{sequence.Name, nil, "optional -> default", "required for optional args"}
+		}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type StaticArgsHaveDefaultsSequenceCheck struct{}
+
+/* Static args must have defaults. */
+func (check StaticArgsHaveDefaultsSequenceCheck) CheckSequence(sequence SequenceSpec) error {
+	for _, arg := range sequence.Args.Static {
+		if arg.Default == nil {
+			return MissingValueError{sequence.Name, nil, "static -> default", "required for static args"}
+		}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type NoDuplicateArgsSequenceCheck struct{}
+
+/* Args should appear once per sequence. */
+func (check NoDuplicateArgsSequenceCheck) CheckSequence(sequence SequenceSpec) error {
+	seen := map[string]bool{}
+	values := map[string]bool{}
+	for _, args := range [][]*ArgSpec{sequence.Args.Required, sequence.Args.Optional, sequence.Args.Static} {
+		for _, arg := range args {
+			if arg.Name != nil && seen[*arg.Name] {
+				values[*arg.Name] = true
+			}
+			seen[*arg.Name] = true
+		}
+	}
+
+	if len(values) > 0 {
+		return DuplicateValueError{sequence.Name, nil, "sequence args -> name", stringSetToArray(values), "sequence args must have unique names within the sequence"}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type HasNodesSequenceCheck struct{}
+
+/* Sequences must contain nodes i.e. are not empty. */
+func (check HasNodesSequenceCheck) CheckSequence(sequence SequenceSpec) error {
+	if len(sequence.Nodes) == 0 {
+		return MissingValueError{sequence.Name, nil, "nodes", "at least one node required"}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type AdminXorOpsSequenceCheck struct{}
+
+/* `admin` and `ops` are mutually exclusive. */
+func (check AdminXorOpsSequenceCheck) CheckSequence(sequence SequenceSpec) error {
+	for _, acl := range sequence.ACL {
+		if acl.Admin && len(acl.Ops) != 0 {
+			return InvalidValueError{sequence.Name, nil, "acl -> admin", []string{"true"}, "admin=false; alternatively, remove ops"}
+		}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type AclsHaveRolesSequenceCheck struct{}
+
+/* ACLs must specify a role. */
+func (check AclsHaveRolesSequenceCheck) CheckSequence(sequence SequenceSpec) error {
+	for _, acl := range sequence.ACL {
+		if acl.Role == "" {
```

Comment:
makes me wonder, should this be a pointer?

Reply:
This is how the old code does it (old as in before I started working on it) so I left it as is, since it implies that the empty string is in fact invalid. Of course, then I can't distinguish between an invalid value (empty string) and actually missing (nil).

This is the same problem as above with `sequence.nodes` and something else (was it `each`?) being empty vs actually missing from the yaml entirely. Is it really useful to have separate checks to distinguish between them? I lean towards no, and either creating some new error that says it's "empty or missing", or just sticking with an existing error and clarifying, e.g. having the `MissingValueError` here say "value cannot be the empty string".